app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js

```diff
@@ -0,0 +1,292 @@
+import {
+  acceptance,
+  exists,
+  query,
+} from "discourse/tests/helpers/qunit-helpers";
+import { click, currentURL, fillIn, visit } from "@ember/test-helpers";
+import pretender from "discourse/tests/helpers/create-pretender";
+import { SECOND_FACTOR_METHODS } from "discourse/models/user";
+import { test } from "qunit";
+
+const { TOTP, BACKUP_CODE, SECURITY_KEY } = SECOND_FACTOR_METHODS;
+
+function setupResponse(response = {}) {
+  const status = response.status || 200;
+  delete response.status;
+  pretender.get("/session/2fa.json", () => {
```

Comment:
I thought this would be a very simple change, but it turned that I needed to restructure this test file. I'm happy to do that, but I just want to double check this first before I continue:

So there are a lots of different states that the `/session/2fa/` page can be in depending on the available 2fa methods for the current user. The page loads the 2fa details of the user from `/session/2fa.json` and based on the server response it determines what to show and what to hide etc. which that means I need pretender to return different responses for the same route to be able test the various states of the page.

Since `needs.pretender` works at the module (I think each `acceptance` block is called a "module"?) level, I'd have to create a module for every different state. For example this test case would need to be split into 3 modules (one for each `setupResponse` call):

https://github.com/discourse/discourse/blob/ecc82cbe4fb55889db285efac888996122996294/app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js#L36-L98

I see most of our acceptance test files have 1 or 2 modules which makes me a little bit unsure if I should create 10+ modules in a single file?

Also just to clarify what you mean by proper clean up: if I use pretender directly like I did here, does it leak my responses to other tests?

Reply:
I'm not sure if pretender leaks currently doing it this way, but I can't guarantee it won't in the future. It's an API we don't want to support.

> which that means I need pretender to return different responses for the same route to be able test the various states of the page.

Did you know pretender can return things conditionally, depending on the input? We commonly use it to reply with fixtures but it's quite powerful. Here's a very simple example:

https://github.com/discourse/discourse/blob/main/app/assets/javascripts/discourse/tests/helpers/create-pretender.js#L146

I think this could help you very much in this situation!

Reply:
> Did you know pretender can return things conditionally, depending on the input?

Yes, I'm aware of that, but the `/session/2fa.json` endpoint takes no parameters; it just responds with the 2fa details of the current user. But from user to user, these details can differ a lot and therefore the 2fa page will look different as well.

I guess I could add a tests-only param that the pretender can check and return the right response for the param value? Something like this:

```javascript
await visit("/session/2fa?pretender_response=response11");
```

Then in the pretender:

```javascript
pretender.get("/session/2fa.json", (request) => {
  return RESPONSES[request.params.pretender_response];
});
```

Or maybe we keep a global variable in the test file that indicates which response the pretender should give, and then each test case changes that global variable to the id of the response it needs?

I like the global variable approach because it can be done without changing anything in non-test files. What's your preference?

Reply:
Ok I've made this change (the one based on the nonce param). @eviltrout can you please take another look (on the pretender bit and the PR in general) and let me know if you're happy with it? thanks!